tempesta_fw/hpack.c

```diff
@@ -0,0 +1,3439 @@
+					goto index;
+				}
+			}
+
+			NEXT_STATE(hp->index
+				   ? HPACK_STATE_INDEXED_NAME_TEXT
+				   : HPACK_STATE_NAME);
+
+			if (!n)
+				goto out;
+
+			if (hp->index) {
+				T_DBG3("%s: decoded index: %lu\n", __func__,
+				       hp->index);
+				goto get_indexed_name;
+			}
+
+			/* Fall through. */
+		}
+		case HPACK_STATE_NAME:
+		{
+			unsigned char c;
+
+			T_DBG("%s: decode header name length...\n", __func__);
+			c = *src++;
+			n--;
+			hp->length = c & 0x7F;
+			if (c & 0x80) {
+				T_DBG("%s: Huffman encoding used for name...\n",
+				      __func__);
+				state |= HPACK_FLAGS_HUFFMAN_NAME;
+			}
+			if (unlikely(hp->length == 0x7F)) {
+				GET_FLEXIBLE(hp->length, HPACK_STATE_NAME_LENGTH);
+			}
+			else if (unlikely(hp->length == 0)) {
+				r = T_DROP;
+				goto out;
+			}
+
+			T_DBG3("%s: name length: %lu\n", __func__, hp->length);
+
+			NEXT_STATE(HPACK_STATE_NAME_TEXT);
+
+			BUFFER_NAME_OPEN(hp->length);
+
+			if (!n)
+				goto out;
+
+			/* Fall through. */
+		}
+		case HPACK_STATE_NAME_TEXT:
+		{
+			bool fnd;
+			unsigned long m_len;
+get_name_text:
+			fnd = false;
+			T_DBG("%s: decode header name...\n", __func__);
+
+			if (state & HPACK_FLAGS_HUFFMAN_NAME)
+				fnd = tfw_hpack_huffman_parse(hp, &src, &n, it);
+
+			if (!fnd) {
+				m_len = min(n, hp->length);
+				if (state & HPACK_FLAGS_HUFFMAN_NAME)
+					HPACK_DECODE_STRING(m_len);
+				else
+					HPACK_COPY_STRING(m_len);
+
+				if (tfw_hpack_hdr_process(req)) {
+					r = T_DROP;
+					goto out;
+				}
+
+				BUFFER_WRITE_COLON();
+			}
+
+			NEXT_STATE(HPACK_STATE_VALUE);
+
+			if (unlikely(!n))
+				goto out;
+
+			goto get_value;
+		}
+		case HPACK_STATE_INDEXED_NAME_TEXT:
+		{
+get_indexed_name:
+			T_DBG("%s: decode indexed (%lu) header name...\n",
+			      __func__, hp->index);
+			WARN_ON_ONCE(!hp->index);
+			hp->entry = tfw_hpack_find_index(&hp->dec_tbl,
+							 hp->index);
+			if (!hp->entry) {
+				r = T_DROP;
+				goto out;
+			}
+
+			BUFFER_NAME_OPEN(hp->entry->name->len);
+			BUFFER_WRITE(hp->entry->name->ptr,
+				     hp->entry->name->len, it);
+
+			if (!STATIC_INDEXED(hp->entry)) {
+				if (tfw_hpack_hdr_process(req)) {
+					r = T_DROP;
+					goto out;
+				}
+			}
+			else {
+				it->hdr_tag = hp->entry->tag;
+			}
+
+			BUFFER_WRITE_COLON();
+
+			NEXT_STATE(HPACK_STATE_VALUE);
+
+			/* Fall through. */
+		}
+		case HPACK_STATE_VALUE:
+		{
+			unsigned char c;
+get_value:
+			T_DBG("%s: decode header value length...\n", __func__);
+			c = *src++;
+			n--;
+			hp->length = c & 0x7F;
+			if (c & 0x80) {
+				T_DBG("%s: Huffman encoding used for value\n",
+				      __func__);
+				state |= HPACK_FLAGS_HUFFMAN_VALUE;
+			}
+			if (unlikely(hp->length == 0x7F))
+				GET_FLEXIBLE(hp->length,
+					     HPACK_STATE_VALUE_LENGTH);
+
+			T_DBG3("%s: value length: %lu\n", __func__, hp->length);
+
+			NEXT_STATE(HPACK_STATE_VALUE_TEXT);
+
+			BUFFER_VAL_OPEN(hp->length);
+
+			if (!n)
+				goto out;
+
+			/* Fall through. */

```

Comment:
Please write a comment what is `nm_node` - that's not clear, especially with special `HPACK_IDX_ST_NM_FOUND` return code.

Reply:
Done.